mono/metadata/sgen-mono-ilgen.c

```diff
@@ -312,13 +312,17 @@ emit_managed_allocator_ilgen (MonoMethodBuilder *mb, gboolean slowpath, gboolean
 		 *
 		 * condition:
 		 *
-		 * bytes <= INT32_MAX - (SGEN_ALLOC_ALIGN - 1)
+		 * bytes <= SIZE_MAX - (SGEN_ALLOC_ALIGN - 1)
 		 *
 		 * therefore:
 		 *
 		 * offsetof (MonoString, chars) + ((len + 1) * 2) <= INT32_MAX - (SGEN_ALLOC_ALIGN - 1)
-		 * len <= (INT32_MAX - (SGEN_ALLOC_ALIGN - 1) - offsetof (MonoString, chars)) / 2 - 1
+		 * len <= (SIZE_MAX - (SGEN_ALLOC_ALIGN - 1) - offsetof (MonoString, chars)) / 2 - 1
+		 * 
+		 * On 64-bit platforms SIZE_MAX is so big that the 32-bit string length can
+		 * never reach the maximum size.
 		 */
```

Comment:
Unrelated: it reminds me there are tests in corefx that expect `new int[int.MaxValue]` to crash because the allowed size is something like int.MaxValue - objsize something like that

Reply:
Also unrelated, the managed code generated for `MONO_BIG_ARRAYS` seems to have the size check wrong.